dirvish_1_3_khl/dirvish-expire.8
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DIRVISH-EXPIRE 8"
.TH DIRVISH-EXPIRE 8 "2005-02-10" "perl v5.8.0" ""
.SH "NAME"
dirvish\-expire \- delete expired dirvish images
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdirvish-expire\fR\fI[\s-1OPTIONS\s0]\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Delete dirvish image trees or whole images that have expired.
.PP
Each image \fBsummary\fR file is checked for the \fBExpire:\fR field. If that
field indicates the image has expired \fBdirvish-expire\fR will delete
that image from the vault.
.PP
By default all subdirectories of all banks will be treated as vaults
and all directories therein except the one named \fBdirvish\fR will be
checked for summary files.
.PP
The removal of an image will have no effect on other images.
.PP
\&\fBDirvish-expire\fR will not delete an image unless it finds at least one
image in that branch that has an intact image tree and \fBStatus:
success\fR in the summary that is not expired.
.SH "OPTIONS"
.IX Header "OPTIONS"
Each option on the command line may be specified any number of times.
Those options that support lists in the config files will accumulate
all of their arguments otherwise each specification will override the
ones before.
.PP
Each option may be unambiguously abbreviated.
.PP
\&\fB\-\-time\fR \fItime_expression\fR
.PP
Execute as though \fItime_expression\fR were the current time.
.PP
\&\fITime_expression\fR is processed by \fB\f(BITime::Parsedate\fB\|(3pm)\fR so relative
time and date strings are permitted. See \fB\f(BITime::Parsedate\fB\|(3pm)\fR for
more details.
.PP
\&\fB\-\-tree\fR
.PP
Only delete the image tree, leave in place the rest of the image
directory with summary, log and any other image administrative files.
.PP
\&\fB\-\-vault\fR \fIvault\fR
.PP
Restrict expiration to the specified \fIvault\fR.
.PP
\&\fB\-\-no\-run\fR
.PP
Don't actually do anything. Just display what would have happened.
.PP
\&\fB\-\-quiet\fR
.PP
Run quietly, only report errors.
.PP
Normally \fBdirvish-expire\fR will report the images deleted.
.SH "EXIT CODES"
.IX Header "EXIT CODES"
To facilitate further automation and integration of \fBdirvish-expire\fR
with other tools \fBdirvish-expire\fR provides rationalised exit codes.
The exit codes are range based. While the code for a specific error may
change from one version to another it will remain within the specified
range. So don't test for specific exit codes but instead test for a
range of values. To the degree possible higher value ranges indicate
more severe errors.
.IP "\fB0\fR" 10
.IX Item "0"
success
.IP "\fB200\-219\fR" 10
.IX Item "200-219"
An error was encountered in loading a configuration file.
.IP "\fB220\-254\fR" 10
.IX Item "220-254"
An error was detected in the configuration.
.IP "\fB255\fR" 10
.IX Item "255"
Incorrect usage.
.SH "FILES"
.IX Header "FILES"
\&\fB/etc/dirvish/master.conf\fR
.PP
alternate master configuration file.
.PP
\&\fB/etc/dirvish.conf\fR
.PP
master configuration file.
.PP
\&\fIbank/vault/image/\fR\fBsummary\fR
.PP
image creation summary.
.PP
\&\fIbank/vault/image/\fR\fBtree\fR
.PP
actual image of source directory tree.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.Vb 2
\& dirvish.conf(5)
\& Time::ParseDate(3pm)
.Ve
.SH "AUTHOR"
.IX Header "AUTHOR"
Dirvish was created by J.W. Schultz of Pegasystems Technologies.
.PP
Dirvish is now maintained by Keith Lofstrom at www.dirvish.org , with
the able assistance of many others.
.SH "BUGS"
.IX Header "BUGS"
\&\fBDirvish-expire\fR will walk the file hierarchy of all banks or the
specified vault looking for summary files. Anything non-dirvish in
there may cause excess file\-walking.


dirvish_1_2/CHANGELOG
2004/02/25           dirvish.conf.5       12.0       Dirvish-1.2        
                     dirvish.pl           12.0       Dirvish-1.2        
                     install.sh           12.0       Dirvish-1.2        
                     loadconfig.pl        12.0       Dirvish-1.2        
                     RELEASE.html         12.0       Dirvish-1.2        
                     TODO.html            12.0       Dirvish-1.2        
                     dirvish-expire.8     12.0       Dirvish-1.2        
                     dirvish-expire.pl    12.0       Dirvish-1.2        
                     dirvish-locate.8     12.0       Dirvish-1.2        
                     dirvish-locate.pl    12.0       Dirvish-1.2        
                     dirvish-runall.8     12.0       Dirvish-1.2        
                     dirvish-runall.pl    12.0       Dirvish-1.2        
                     dirvish.8            12.0       Dirvish-1.2        
                     COPYING              12.0       Dirvish-1.2        
                     FAQ.html             12.0       Dirvish-1.2        
                     INSTALL              12.0       Dirvish-1.2        
                     RELEASE.html         11.7                         
                     FAQ.html             11.8                         
Preparing to tag 1.2.0

Note license change and prep for 1.2

Update license note.

2004/02/23           dirvish.pl           11.27                        
                     install.sh           11.3                         
Set --delete on rsync so retries will delete deleted files.
					(Thanks to Nathan Hunsperger)

Allow installs to existing install dir with missing subdirs.

2004/01/20           FAQ.html             11.7                         
Removed banner reference.

2004/01/19           FAQ.html             11.6                         
                     dirvish.pl           11.26                        
Reordered chmod operations on image.
					thanks to Nathan Hunsperger

2003/12/21           dirvish.pl           11.25                        
Better handling of vanished files on sender.

2003/12/17           TODO.html            11.16                        
file security check added to wish list

2003/12/15           dirvish.pl           11.24                        
Added recognition for vanishing (unexpected EOF) files on
client.  This happens due to a file being deleted or
truncated between file scan and send.

2003/12/11           FAQ.html             11.5                         
                                          11.4                         
Fixed html tag typo				-- Frank de Jong

2003/12/08           dirvish.pl           11.23                        
Fix permissions of rsync-error file.
					Thanks to Nathan Hunsperger

2003/12/06           INSTALL              11.4                         
Change to reflect rsync's 2.5.7 security release and plan
for post 2.5.6 CVS HEAD to be 2.6.0

2003/12/03           dirvish.conf.5       11.12                        
fixed some references to expire-rules (wrong plural) and
expire_rules (underscore)

2003/11/26           dirvish.conf.5       11.11                        
                     dirvish.pl           11.22                        
                                          11.21                        
                     TODO.html            11.15                        
Added a devices option (default yes) to allow excluding
devices from backups.

Fixed problems with expire-default and expire rules that
assume never.

2003/10/25           dirvish-expire.pl    11.12                        
Allow a vault named dirvish.
A vault named "tree" will still be skipped.

2003/09/17           dirvish.conf.5       11.10                        
fix typo

2003/09/14           dirvish.pl           11.20                        
                     loadconfig.pl        11.13                        
                     TODO.html            11.14                        
Check to be sure that pre and post scripts are specified
correctly (single value) in config file.

Added whitespace after seppuku_prefix

2003/09/01           TODO.html            11.13                        
                     dirvish-expire.pl    11.11                        
fixed dirvish-expire --vault so it works

Still need major rework.

2003/08/30           TODO.html            11.12                        
2003/08/29           dirvish.pl           11.19                        
                                          11.18                        
                     loadconfig.pl        11.12                        
                     TODO.html            11.11                        
                     INSTALL              11.3                         
                     FAQ.html             11.3                         
                     dirvish.conf.5       11.9                         
Fix seppuku prefix so it is $vault:$branch:$image

Added seppuku prefix variable so error messages will be
suitably marked with the vault:

Better reflect filesystem requirements in INSTALL and FAQ.

2003/08/03           dirvish-expire.pl    11.10                        
                     dirvish-locate.pl    11.9                         
                     dirvish-runall.pl    11.9                         
                     dirvish.pl           11.17                        
                     loadconfig.pl        11.11                        
                     COPYING              11.5                         
                                          11.4                         
                     dirvish-expire.pl    11.9                         
                     dirvish-locate.pl    11.8                         
                     dirvish-runall.pl    11.8                         
                     dirvish.pl           11.16                        
                     loadconfig.pl        11.10                        

dirvish_1_2/dirvish-expire.pl


#       $Id: dirvish-expire.pl,v 12.0 2004/02/25 02:42:14 jw Exp $  $Name: Dirvish-1_2 $

$VERSION = ('$Name: Dirvish-1_2 $' =~ /Dirvish/i)
	? ('$Name: Dirvish-1_2 $' =~ m/^.*:\s+dirvish-(.*)\s*\$$/i)[0]
	: '1.1.2 patch' . ('$Id: dirvish-expire.pl,v 12.0 2004/02/25 02:42:14 jw Exp $'
		=~ m/^.*,v(.*:\d\d)\s.*$/)[0];
$VERSION =~ s/_/./g;


#########################################################################
#                                                         		#
#	Copyright 2002 and $Date: 2004/02/25 02:42:14 $
#                         Pegasystems Technologies and J.W. Schultz 	#
#                                                         		#
#	Licensed under the Open Software License version 2.0		#
#                                                         		#
#	This program is free software; you can redistribute it		#
#	and/or modify it under the terms of the Open Software		#
#	License, version 2.0 by Lauwrence E. Rosen.			#
#                                                         		#
#	This program is distributed in the hope that it will be		#
#	useful, but WITHOUT ANY WARRANTY; without even the implied	#
#	warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR		#
#	PURPOSE.  See the Open Software License for details.		#
#                                                         		#
#########################################################################

use Time::ParseDate;
use POSIX qw(strftime);
use File::Find;
use Getopt::Long;

sub loadconfig;
sub check_expire;
sub findop;
sub imsort;
sub seppuku;

sub usage
{
	my $message = shift(@_);

	length($message) and print STDERR $message, "\n\n";

	print STDERR <<EOUSAGE;
USAGE
	dirvish.expire OPTIONS
	
OPTIONS
	--time date_expression
	--[no]tree
	--vault vault_name
	--no-run
	--quiet
EOUSAGE

	exit 255;
}

$Options = 
{ 
	help		=> \&usage,
	version		=> sub {
			print STDERR "dirvish version $VERSION\n";
			exit(0);
		},
};

if ($CONFDIR =~ /dirvish$/ && -f "$CONFDIR.conf")
{
	loadconfig(undef, "$CONFDIR.conf", $Options);
}
elsif (-f "$CONFDIR/master.conf")
{
	loadconfig(undef, "$CONFDIR/master.conf", $Options);
}
elsif (-f "$CONFDIR/dirvish.conf")
{
	seppuku 250, <<EOERR;
ERROR: no master configuration file.
	An old $CONFDIR/dirvish.conf file found.
	Please read the dirvish release notes.
EOERR
}
else
{
	seppuku 251, "ERROR: no global configuration file";
}

GetOptions($Options, qw(
	quiet!
	vault=s
	time=s
	tree!
	no-run|dry-run
	version
	help|?
	)) or usage;

ref($$Options{vault}) and $$Options{vault} = undef;

$$Options{time} and $expire_time = parsedate($$Options{time});
$expire_time ||= time;


if ($$Options{vault})
{
	my $b;
	my $bank;
	for $b (@{$$Options{bank}})
	{
		if (-d "$b/$$Options{vault}")
		{
			$bank = $b;
			last;
		}
	}
	$bank or seppuku 252, "Cannot find vault $$Options{vault}";
	# Follow symlinks                                                                                                                                                                                      
        find( {wanted => \&findop, follow => 1}, join('/', $bank, $$Options{vault}));
} else {
	for $bank (@{$$Options{bank}})
	{
		find(\&findop, $bank);
	}
}

scalar(@expires) or exit 0;

if (!$$Options{quiet})
{
	printf "Expiring images as of %s\n",
		strftime('%Y-%m-%d %H:%M:%S', localtime($expire_time));
	$$Options{vault} and printf "Restricted to vault %s\n",
		$$Options{vault};
	print "\n";

	printf "%-15s %-15s %-16.16s  %s\n",
		qw(VAULT:BRANCH IMAGE CREATED EXPIRED);
}

for $expire (sort(imsort @expires))
{
	my ($created, $expired);
	($created = $$expire{created}) =~ s/:\d\d$//;
	($expired = $$expire{expire}) =~ s/:\d\d$//;
	
	if (!$unexpired{$$expire{vault}}{$$expire{branch}})
	{
		printf "cannot expire %s:%s:%s No unexpired good images\n",
			$$expire{vault},
			$$expire{branch},
			$$expire{image};
		$$expire{status} =~ /^success/
			and ++$unexpired{$$expire{vault}}{$$expire{branch}};
		# By virtue of the sort order this will be the newest 
		# image so that older ones can be expired.
		next;
	}
	$$Options{quiet} or printf "%-15s %-15s %-16.16s  %s\n",
		$$expire{vault} . ':' .  $$expire{branch},
		$$expire{image},
		$created,
		$expired;

	$$Options{'no-run'} and next;

	system("rm -rf $$expire{path}/tree");
	$$Options{tree} and next;

	system("rm -rf $$expire{path}");
}

exit 0;

sub check_expire
{
	my ($summary, $expire_time) = @_;

	my ($expire, $etime, $path);

	$expire = $$summary{Expire};
	$expire =~ s/^.*==\s+//;
	$expire or return 0;
	$expire =~ /never/i and return 0;

	$etime = parsedate($expire);

	if (!$etime)
	{
		print STDERR "$File::Find::dir: invalid expiration time $$summary{expire}\n";
		return -1;
	}
	$etime > $expire_time and return 0;

	return 1;
}

sub findop
{
	if ($_ eq 'tree')
	{
		$File::Find::prune = 1;
		return 0;
	}
	if ($_ eq 'summary')
	{
		my $summary;
		my ($etime, $path);

		$path = $File::Find::dir;

		$summary = loadconfig('R', $File::Find::name);
		$status = check_expire($summary, $expire_time);
		
		$status < 0 and return;

		$$summary{vault} && $$summary{branch} && $$summary{Image}
			or return;

		if ($status == 0)
		{
			$$summary{Status} =~ /^success/ && -d ($path . '/tree')
				and ++$unexpired{$$summary{vault}}{$$summary{branch}};
			return;
		}

		-d ($path . ($$Options{tree} ? '/tree': undef)) or return;

		push (@expires, {
				vault	=> $$summary{vault},
				branch	=> $$summary{branch},
				client	=> $$summary{client},
				tree	=> $$summary{tree},
				image	=> $$summary{Image},
				created	=> $$summary{'Backup-complete'},
				expire	=> $$summary{Expire},
				status	=> $$summary{Status},
				path	=> $path,
			}
		);
	}
}

## WARNING:  don't mess with the sort order, it is needed so that if
## WARNING:  all images are expired the newest will be retained.
sub imsort
{
	$$a{vault} cmp $$b{vault}
	|| $$a{branch} cmp $$b{branch}
	|| $$a{created} cmp $$b{created};
}

